wpsc-components/merchant-core-v3/gateways/paypal-express-checkout.php

```diff
@@ -7,37 +7,178 @@
-			'currency'         => $this->get_currency_code(),
-			'test'             => (bool) $this->setting->get( 'sandbox_mode' ),
-			'address_override' => 1,
-			'solution_type'    => 'mark',
-			'cart_logo'        => $this->setting->get( 'cart_logo' ),
-			'cart_border'      => $this->setting->get( 'cart_border' ),
+		$this->gateway = new PHP_Merchant_Paypal_Express_Checkout( $options );		
+
+		if ( ! $child ) {
+			$this->title = __( 'PayPal Express Checkout 3.0', 'wpsc' );
+			$this->gateway->set_options( array(
+				'api_username'     => $this->setting->get( 'api_username' ),
+				'api_password'     => $this->setting->get( 'api_password' ),
+				'api_signature'    => $this->setting->get( 'api_signature' ),
+				'cancel_url'       => $this->get_shopping_cart_payment_url(),
+				'currency'         => $this->get_currency_code(),
+				'test'             => (bool) $this->setting->get( 'sandbox_mode' ),
+				'address_override' => 1,
+				'solution_type'    => 'mark',
+				'cart_logo'        => $this->setting->get( 'cart_logo' ),
+				'cart_border'      => $this->setting->get( 'cart_border' ),
+			) );
+
+			// Express Checkout Button	
+			add_action( 'wpsc_cart_item_table_after', array( &$this, 'add_ecs_button' ), 0, 10 );
+		}
+	}
+
+    /**
+     * Insert the ExpessCheckout Shortcut Button
+     *
+     * @return void
+     */
+	public function add_ecs_button() {
+		if ( _wpsc_get_current_controller_name() === 'cart' ) {	
+			$url = $this->get_shortcut_url();
+			echo '<a href="'. $url .'"><img src="https://www.paypalobjects.com/webstatic/en_US/i/buttons/checkout-logo-large.png" alt="Check out with PayPal" /></a>';
+		}
+	}
+
+    /**
+     * Return the ExpressCheckout Shortcut redirection URL
+     *
+     * @return void
+     */
+	public function get_shortcut_url() {
+		$location = add_query_arg( array(
+			'payment_gateway'          => 'paypal-express-checkout',
+			'payment_gateway_callback' => 'shortcut_process',
+		), home_url( 'index.php' ) );
+
+		return apply_filters( 'wpsc_paypal_express_checkout_shortcut_url', $location );
+	}
+
+    /**
+     * ExpressCheckout Shortcut Callback
+     *
+     * @return void
+     */
+	public function callback_shortcut_process() {
+        if ( ! isset( $_GET['payment_gateway'] ) ) {
+            return;
+        }
+        $payment_gateway = $_GET['payment_gateway'];
+        
+		global $wpsc_cart;
+		//	Create a new PurchaseLog Object 
+		$purchase_log = new WPSC_Purchase_Log();	
+
+		// Create a Sessionid
+		$sessionid = ( mt_rand( 100, 999 ) . time() );
+		wpsc_update_customer_meta( 'checkout_session_id', $sessionid );
+		$purchase_log->set( array(
+			'user_ID'        => wpsc_get_current_customer_id(),
+			'date'           => time(),
+			'plugin_version' => WPSC_VERSION,
+			'statusno'       => '0',
+			'sessionid'      => $sessionid,
+        ) );
+
+        // Set a Transient for EC Shortcut
+        set_transient( 'ecs-' . $sessionid, true, 60 * 60 );
```

Comment:
Is using a transient the best thing here?  Transients can be deleted or unavailable at any time.